Quarantine and cleanup idle associations, #24972

patriknw opened the pull request:
Some first improvements related to #24972

The reproducer app can be run with:
```
sbt "akka-remote/test:runMain akka.remote.artery.tcp.LeakyServer"
sbt "akka-remote/test:runMain akka.remote.artery.tcp.PingClient"
```

or corresponding with cluster:
```
sbt "akka-cluster/test:runMain akka.cluster.LeakyServer"
sbt "akka-cluster/test:runMain akka.cluster.PingClient"
```

patriknw committed:
Quarantine and cleanup idle associations, #24972

* fix NPE in shutdownTransport
  * perhaps because shutdown before started
  * system.dispatcher is used in other places of the shutdown
* improve logging of compression advertisment progress
* adjust RestartFlow.withBackoff parameters
* quarantine after ActorSystemTerminating signal
  (will cleanup compressions)
* Quarantine idle associations
  * liveness checks by sending extra HandshakeReq and update the
    lastUsed when reply received
  * concervative default value to survive network partition, in
    case no other messages are sent
* Adjust logging and QuarantinedEvent for harmless quarantine
  * Harmless if it was via the shutdown signal or cluster leaving

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ArteryTcpTransport.scala:
```diff
@@ -152,19 +152,13 @@ private[remote] class ArteryTcpTransport(_system: ExtendedActorSystem, _provider
-        // are buffered and resent from the outer SystemMessageDelivery stage.
-        RestartFlow.withBackoff[ByteString, ByteString](
-          settings.Advanced.OutboundRestartBackoff,
-          settings.Advanced.GiveUpSystemMessageAfter, 0.1)(flowFactory)
```
this max of `GiveUpSystemMessageAfter` must have been a thinko, no reason to exponentially increase the backoff that much

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -812,13 +812,20 @@ private[remote] abstract class ArteryTransport(_system: ExtendedActorSystem, _pr
+              if (inControlStream)
+                system.scheduler.scheduleOnce(settings.Advanced.ShutdownFlushTimeout) {
+                  if (!isShutdown)
+                    quarantine(from.address, Some(from.uid), "ActorSystem terminated")
```
This improves the situation for the normal ActorSystem shutdown since we cleanup many things when it's quarantined. It will not capture the case when the remote `kill -9`.

For a cluster the situation much better because we quarantine removed members.

patriknw replied:
I have confirmed that this is also solving the leak of `ActorRef`s that was reported in https://github.com/akka/akka/issues/24973

I'll look into the `kill -9` case for that now.

patriknw replied:
To clarify, some `ActorRef` remains in the `ActorRefResolveCache`, but its size is bounded to 1024. I have verified that this bound works in practice by setting it to a lower value.

luben commented on akka-remote/src/main/scala/akka/remote/artery/tcp/ArteryTcpTransport.scala:
```diff
@@ -152,19 +152,13 @@ private[remote] class ArteryTcpTransport(_system: ExtendedActorSystem, _provider
-          settings.Advanced.OutboundRestartBackoff * 5, 0.1, maxRestarts = 3)(flowFactory)
-      }
-
+      val maxRestarts = if (streamId == ControlStreamId) Int.MaxValue else 3
```
If we gave up the messages after "settings.Advanced.GiveUpSystemMessageAfter", why we retry infinitely to re-establish the control association? Or you are giving up in some other part of the code?

patriknw replied:
GiveUpSystemMessageAfter is used by the SystemMessageDelivery stage to resend system messages like Watch and DeathWatchNotifications. Using it here was wrong.

Note that the `maxBackoff` parameter to `RestartFlow.withBackoff` is not a give up, it's the upper bound of the backoff interval.

patriknw commented:
> Rerun the tests with the proposed patch. It fixes the both leaks in the normal exit case but the re-connection is still happening if we kill -9 the client: I just made client actor not stop itself so that I have a chance to kill it manually.

Thanks for confirming. My plan is to add another tear down mechanism that quarantines associations that have been idle (no messages sent) for a configured timeout. That is not really needed for clustered applications but if plain remoting is used that can be a way to clean up such crashed connections.

luben commented:
Another question, in the normal case where the client is gracefully shut down, I still see the connection quarantined, e.g:
```
[DEBUG] [04/25/2018 20:56:31.754] [Server-akka.actor.default-dispatcher-9] [akka://Server@127.0.1.1:2552/system/IO-TCP/selectors/$a/2] Closing connection due to IO error java.io.IOException: Connection reset by peer
[DEBUG] [04/25/2018 20:56:31.757] [Server-akka.actor.default-dispatcher-2] [akka://Server@127.0.1.1:2552/system/IO-TCP/selectors/$a/1] Closing connection due to IO error java.io.IOException: Connection reset by peer
[DEBUG] [04/25/2018 20:56:31.803] [Server-akka.remote.default-remote-dispatcher-4] [akka.stream.Log(akka://Server/system/StreamSupervisor-0)] [outbound connection to [akka://Client@127.0.1.1:34517], message stream] Upstream finished.
[DEBUG] [04/25/2018 20:56:31.803] [Server-akka.remote.default-remote-dispatcher-4] [RestartWithBackoffFlow(akka://Server)] Restarting graph due to finished upstream
[DEBUG] [04/25/2018 20:56:31.804] [Server-akka.remote.default-remote-dispatcher-5] [akka.stream.Log(akka://Server/system/StreamSupervisor-0)] [outbound connection to [akka://Client@127.0.1.1:34517], control stream] Upstream finished.
[DEBUG] [04/25/2018 20:56:31.804] [Server-akka.remote.default-remote-dispatcher-5] [RestartWithBackoffFlow(akka://Server)] Restarting graph due to finished upstream
[DEBUG] [04/25/2018 20:56:31.820] [Server-akka.remote.default-remote-dispatcher-5] [RestartWithBackoffFlow(akka://Server)] Restarting graph in 10560401223 nanoseconds
[DEBUG] [04/25/2018 20:56:31.833] [Server-akka.remote.default-remote-dispatcher-4] [RestartWithBackoffFlow(akka://Server)] Restarting graph in 10401775745 nanoseconds
[WARN] [04/25/2018 20:56:32.747] [Server-akka.remote.default-remote-dispatcher-4] [akka.remote.artery.Association(akka://Server)] Association to [akka://Client@127.0.1.1:34517] with UID [-8390001983367142586] is irrecoverably failed. UID is now quarantined and all messages to this UID will be delivered to dead letters. Remote actorsystem must be restarted to recover from this situation. Reason: ActorSystem terminated
[DEBUG] [04/25/2018 20:56:32.751] [Server-akka.remote.default-remote-dispatcher-4] [akka.remote.artery.Association(akka://Server)] Send control message [akka.remote.artery.Quarantined] to quarantined [akka://Client@127.0.1.1:34517]

[DEBUG] [04/25/2018 20:56:35.775] [Server-akka.remote.default-remote-dispatcher-5] [akka.remote.artery.Association(akka://Server)] Outbound control stream to [akka://Client@127.0.1.1:34517] was quarantined and stopped. It will be restarted if used again.
[DEBUG] [04/25/2018 20:56:35.775] [Server-akka.remote.default-remote-dispatcher-4] [akka.remote.artery.Association(akka://Server)] Outbound message stream to [akka://Client@127.0.1.1:34517] was quarantined and stopped. It will be restarted if used again.

```

My question is, why quarantine the association when we are sure that the remote was properly shut down? I tend to believe that quarantining is something bad, signal for things gone wrong, but now it will happen all the time in the normal course of events.

patriknw commented:
> quarantining is something bad

That's a good point. Quarantine is not something bad here. It means that we can't send messages to that specific incarnation hostname:port/uid, and we don't want that anyway since it's shutdown.

I'll revise the log message. Best would be to have a different message for this graceful shutdown scenario (same for cluster removal via leaving).

patriknw commented:
@luben I have added additional cleanup of the dead (kill -9) associations. Configured via `quarantine-idle-outbound-after (see documentation in reference.conf). It's commit d61a0a2 if you want to look into the details around that specific case.

I have also adjusted the logging for the "harmless" quarantine, and for them I'm publishing another event, `HarmessQuarantinedEvent`.

Please give it another try. See latest AkkaLeak.scala for config that is decent for testing.

> when quarantining, aren't we also keeping a tombstone so that we stop accepting new associations/messages from the remote. What is the purpose of it if we know that the remote has indeed terminated gracefully - it can't be resurrected?

I don't want to introduce "another" state for that case. There can be race conditions with messages to the next incarnation.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressions.scala:
```diff
@@ -354,15 +355,17 @@ private[remote] abstract class InboundCompression[T >: Null](
-        log.debug("Confirmed compression table version [{}] for originUid [{}]", tableVersion, originUid)
+        log.debug(
+          "{} compression table version [{}] for originUid [{}]",
+          if (gaveUp) "Gave up" else "Confirmed", tableVersion, originUid)
```
This is on the destination side, which builds up the tables and announce them to the sending side. This only means that the sending side have not seen the new table and will therefore not use it. The destination side keeps several tables around so that sender can still use old table.

johanandren replied:
Ah, thanks I missed out the `Inbound` part.

luben commented:
@patriknw, it works as expected. Thanks for the fast fix.

patriknw commented:
Rebased, ready for final review. The summary (commit message):

* fix NPE in shutdownTransport
  * perhaps because shutdown before started
  * system.dispatcher is used in other places of the shutdown
* improve logging of compression advertisment progress
* adjust RestartFlow.withBackoff parameters
* quarantine after ActorSystemTerminating signal
  (will cleanup compressions)
* Quarantine idle associations
  * liveness checks by sending extra HandshakeReq and update the
    lastUsed when reply received
  * concervative default value to survive network partition, in
    case no other messages are sent
* Adjust logging and QuarantinedEvent for harmless quarantine
  * Harmless if it was via the shutdown signal or cluster leaving

patriknw commented:
Refs #24972

patriknw commented:
Refs https://github.com/akka/akka/issues/24973

johanandren reviewed (approved):
LGTM, but quite massive so not sure I understand all consequences

chbatey commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -988,6 +989,23 @@ akka {
+        # This will cleanup association to crashed systems that didn't announce their
+        # termination.
+        # The value should be longer than the length of a network partition that you
+        # need to survive. The value must also be greater than stop-idle-outbound-after.
```
I think this needs to be 10x greater than the stop-idle-outbound-after otherwise this hand shake will keep the outbound lanes from every been shut down.

patriknw replied:
Are you sure? it was long time since I worked on this, but I think the low frequency handshakes (quarantine-idle-outbound-after / 10) don't touch the timestamp when they are sent, but when the reply is received.

Even though that could keep the connection alive if the systems are alive, and replies, I don't see that as a big problem. The connection will be woken up at quarantine-idle-outbound-after / 10 no matter. Low values of quarantine-idle-outbound-after are not recommended, but I don't think we have to enforce some value.

Perhaps you don't want to stop idle at all, but still want this feature.

chbatey replied:
Yes I accidently came across this when trying different versions of each timeout.

Ah yes i see if we put that limit in we would be disabling users from using this feature but not stopping idle via `stop-idle-outbound-after`. 

I still find it surprising in the case that I want to clean up TCP connections if there are no user messages and want this new feature I can accidently disable the `stop-idle-outbound-after` via the value of `quarantine-idle-outbound-after`

patriknw replied:
Ok, we can add that 10x requirement, but then also support `off` as a possible value in `stop-idle-outbound-after`.

patriknw replied:
or just add a note to the reference.conf (not worth putting too much effort into this)

chbatey replied:
:+1: note is fine

chbatey reviewed (approved):
LGTM

A few minor comments, one for how the config values depend on each other that i think we should add another require for.

I went through each timeout and tried them. I'm glad they are in advanced as it took me a while to grok them all. I think the defaults are sensible.

Might want to make a minor update to https://github.com/akka/akka/blame/master/akka-docs/src/main/paradox/remoting-artery.md#L549 for this new harmless Quarantine?

One thing that came to mind but isn't specifically for this PR is that we log warnings at the streams TCP layer in artery even when another ActorSystem shuts down gracefully.

patriknw commented:
BTW, I ran this PR in repeat job for a few days

patriknw commented on akka-remote/src/main/scala/akka/remote/RemotingLifecycleEvent.scala:
```diff
@@ -110,6 +110,14 @@ final case class QuarantinedEvent(address: Address, longUid: Long) extends Remot
 }
 
+@SerialVersionUID(1L)
+final case class HarmlessQuarantinedEvent(uniqueAddress: UniqueAddress, reason: String) extends RemotingLifecycleEvent {
```
Yes, it's when cluster member is removed via leaving/exiting, or if the other actor system signaled that it was shutdown. Ok. I can rename it to `GracefulShutdownQuarantineEvent`.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -530,21 +538,22 @@ private[remote] class Association(
   }
 
   private def abortQuarantined(): Unit = {
+    cancelIdleTimer()
```
I think the interesting part of this is logged slightly later as "was quarantined and stopped. It will be restarted if used again"

patriknw commented:
updated with some of the feedback, see f58049e

chbatey commented:
Probably worth adding some notes to the reference docs but maybe not needed as they are all "advanced" settings.

patriknw commented:
> Probably worth adding some notes to the reference docs but maybe not needed as they are all "advanced" settings.

Yeah, I don't want to advertise this too much. I'd prefer that default values are used. It's mostly because we don't like having leaks in our software.
